oxenmq/bt_producer.h

```diff
@@ -0,0 +1,306 @@
+
+    return pos;
+}
+#endif
+
+
+/// Class that allows you to build a bt-encoded list manually, without copying or allocating memory.
+/// This is essentially the reverse of bt_list_consumer: where it lets you stream-parse a buffer,
+/// this class lets you build directly into a buffer that you own.
+///
+/// Out-of-buffer-space errors throw 
+class bt_list_producer {
+    friend class bt_dict_producer;
+
+    // Our pointers to the next write position and the past-the-end pointer of the buffer.
+    using buf_span = std::pair<char*, char*>;
+    // Our data is a begin/end pointer pair for the root list, or a pointer to our parent if a
+    // sublist.
+    std::variant<buf_span, bt_list_producer*, bt_dict_producer*> data;
+    // Reference to the write buffer; this is simply a reference to the value inside `data` for the
+    // root element, and a pointer to the root's value for sublists/subdicts.
+    buf_span& buffer;
+    // True indicates we have an open child list/dict
+    bool has_child = false;
+    // The range that contains this currently serialized value; `from` equals wherever the `l` was
+    // written that started this list and `to` is one past the `e` that ends it.  Note that `to`
+    // will always be ahead of `buf_span.first` because we always write the `e`s to close open lists
+    // but these `e`s don't advance the write position (they will be overwritten if we append).
+    const char* const from;
+    const char* to;
+
+    // Sublist constructors
+    bt_list_producer(bt_list_producer* parent, std::string_view prefix = "l"sv);
+    bt_list_producer(bt_dict_producer* parent, std::string_view prefix = "l"sv);
+
+    // Does the actual appending to the buffer, and throwing if we'd overrun.  If advance is false
+    // then we append without moving the buffer pointer (primarily when we append intermediate `e`s
+    // that we will overwrite if more data is added).  This means that the next write will overwrite
+    // whatever was previously written by an `advance=false` call.
+    void buffer_append(std::string_view d, bool advance = true);
+
+    // Appends the 'e's into the buffer to close off open sublists/dicts *without* advancing the
+    // buffer position; we do this after each append so that the buffer always contains valid
+    // encoded data, even while we are still appending to it, and so that appending something raises
+    // a length_error if appending it would not leave enough space for the required e's to close the
+    // open list(s)/dict(s).
+    void append_intermediate_ends(size_t count = 1);
+
+    // Writes an integer to the given buffer; returns the one-past-the-data pointer.  Up to 20 bytes
+    // will be written and must be available in buf.  Used for both string and integer
+    // serialization.
+    template <typename IntType>
+    char* write_integer(IntType val, char* buf) {
+        static_assert(sizeof(IntType) <= 64);
+
+#ifndef OXENMQ_APPLE_TO_CHARS_WORKAROUND
```

Comment:
i kind of like doing `#ifdef WHATEVER` instead of `#ifndef WHATEVER` as the latter feels upside down.

Reply:
In this case, I did it this way so that the "good" implementation is the first one you see, followed by the workaround.